File: python/solutions/p003_largest_prime.py
#!/usr/bin/env python3 -OO

"""Largest prime factor
Problem 3
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
"""

import math

def run(num):
    factors = []
    for idx in range(1, int(math.sqrt(num))):
        if num % idx == 0:
            if is_prime(idx):
                factors.append(idx)
            if is_prime(num/idx):
                factors.append(num/idx)
    return max(factors)


def is_prime(num):
    if num <= 1:
        return False
    elif num <= 3:
        return True
    elif num % 2 == 0 or num % 3 == 0:
        return False
    i = 5
    while i * i <= num:
        if num % i == 0 or num % (i + 2) == 0:
            return False
        i += 6
    return True

File: ruby/solutions/p003_largest_prime.rb
#!/usr/bin/env ruby -w

require 'prime'

module P003
  def self.run num
    factors = []
    (2..Math.sqrt(num).to_i).each do |i|
      if num % i == 0
        if Prime.prime?(i)
          factors << i
        end
        if Prime.prime?(i/num)
          factors << i/num
        end
      end
    end

    factors.max
  end

  # alternative solution one line solution
  def self.run2 num
    (2..num).select { |i| num % i == 0 and Prime.prime?(i) }.max
  end
end
